Make web::Path innards public

IriaSomobu opened the pull request:
## PR Type

Other?

## PR Checklist

- [x] Tests for the changes have been added / updated.
- [x] ~Documentation comments have been added / updated~
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt.
- [x] (Team) Label with affected crates and semver status.

## Overview

This change allows us to do something like this:
```rust
async fn delete_channel_message(
    web::Path((channel_id, message_id)): web::Path<(Snowflake, Snowflake)>,
) {
  // here channel_id and message_id are Snowflakes,
  // and not web::Path<Snowflake>
}
```

Dunno why other web types are public and this is not.

IriaSomobu committed:
Make web::Path innards public

So we can do something like this:
```
async fn delete_channel_message(
    web::Path((channel_id, message_id)): web::Path<(Snowflake, Snowflake)>,
) {
  // here channel_id and message_id are Snowflakes,
  // and not web::Path<Snowflake>
}
```

Dunno why other web types are public and this is not.

robjtede commented:
This field was made private for v4.0 but later it was realized this was the wrong trade-off. Making it public again is certainly planned for v5.0.

This is a breaking change due to the Deref impl, so it should also be removed in this PR.

For now, you can use the [Path type from actix-web-lab](https://docs.rs/actix-web-lab/latest/actix_web_lab/extract/struct.Path.html) to get destructuring.

IriaSomobu commented:
> This is a breaking change due to the Deref impl, so it should also be removed in this PR.

Could you elaborate please?

robjtede commented:
The thing we were seeing in v3.x was users reporting confusion using this type without destructuring because the Deref impl allowed access to tuple elements like `path.1` but then `path.0.1` also worked so it was weird. Anyway, making this public will make the `path.0` case ambiguous to the compiler in certain cases, making this a breaking change.

IriaSomobu commented:
Okey, I get it. `path.0.1` really looks funny. Gimme 5min -- I'll remove that deref.

robjtede commented:
No rush. Like I say, this will need to wait for v5.0.

IriaSomobu committed:
Drop Deref and DerefMut on web::Path

@robjtede: Deref impl [with public Path contents] allowed access to tuple elements like path.1 but then path.0.1 also worked so it was weird.
